### PR TITLE
fix: set timeout for delete work

### PR DESCRIPTION
**Pull request by ycyaoxdu**
Signed-off-by: ycyaoxdu <yaoyuchen0626@163.com>

#159 set a timeout for deleting work.

### CODE REVIEW

**Review by qiujian16 — COMMENTED** [review]
I think we should also add a `--force` flag so we delete the work at first, and if it is not removed, remove its finalizer to force remove it.

**Comment by ycyaoxdu** [comment]
```
root@yuchen clusteradm [fix] $ clusteradm get clusters
NAME       ACCEPTED   AVAILABLE   CLUSTERSET   CPU   MEMORY       KUBERENETES VERSION
cluster2   true       Unknown                  8     16185152Ki   v1.21.1
root@yuchen clusteradm [fix] $ clusteradm get works --cluster cluster2
NAME    CLUSTER    NUMBER OF MANIFESTS   APPLIED   AVAILABLE
work2   cluster2   1                     True      True
root@yuchen clusteradm [fix] $ clusteradm delete work work2 --cluster cluster2 
Error: delete work work2 timeout, failed to delete
root@yuchen clusteradm [fix] $ clusteradm delete work work2 --cluster cluster2 --force
work work2 in cluster cluster2 is deleted
root@yuchen clusteradm [fix] $ clusteradm get works --cluster cluster2

```

File: pkg/cmd/delete/work/exec.go [hunk 1]
```diff
@@ -51,21 +52,53 @@ func (o *Options) run() error {
...
+
+	if o.Force {
+		work.ObjectMeta.Finalizers = work.ObjectMeta.Finalizers[:0]
+		_, err = workClient.WorkV1().ManifestWorks(o.Cluster).Update(context.TODO(), work, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	err = workClient.WorkV1().ManifestWorks(o.Cluster).Delete(context.TODO(), o.Workname, metav1.DeleteOptions{})
```
Comment (qiujian16) [inline pkg/cmd/delete/work/exec.go:94]: run delete at first then remove finalizer, otherwise there is chance that finalizer is added back

Comment (ycyaoxdu) [reply pkg/cmd/delete/work/exec.go]: done

File: pkg/cmd/delete/work/exec.go [hunk 2]
```diff
@@ -51,21 +52,53 @@ func (o *Options) run() error {
...
+
+	}(errChannel)
+
+	err = workClient.WorkV1().ManifestWorks(o.Cluster).Delete(context.TODO(), o.Workname, metav1.DeleteOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if o.Force {
+		work.ObjectMeta.Finalizers = work.ObjectMeta.Finalizers[:0]
```
Comment (qiujian16) [inline pkg/cmd/delete/work/exec.go:92]: nit you probably want to get again here.

File: pkg/cmd/delete/work/exec.go [hunk 3]
```diff
@@ -51,21 +52,64 @@ func (o *Options) run() error {
...
+
+	err = workClient.WorkV1().ManifestWorks(o.Cluster).Delete(context.TODO(), o.Workname, metav1.DeleteOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if o.Force {
+		// check whether work is already deleted, if not, remove the finalizer
+		work, err = workClient.WorkV1().ManifestWorks(o.Cluster).Get(context.TODO(), o.Workname, metav1.GetOptions{})
+		if err != nil {
```
Comment (qiujian16) [inline pkg/cmd/delete/work/exec.go:94]: I would do 

```
if errors.IsNotFound(err) {
	fmt.Fprintf(o.Streams.Out, "work %s is deleted\n", o.Workname)
	return nil
}

if err != nil {
    return err
}

if len(work.Finalizers) == 0 {
    return nil
}

```

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**